.github/workflows/test-CI.yml

```diff
@@ -0,0 +1,166 @@
+name: Test Library CI
+
+on:
+  # Trigger the workflow on push to the specific branch
+  push:
+    branches:
+      - dev
+      - main
+      - CAD-22795-restore-changes # Temp
+
+  # Trigger the workflow on pull requests targeting the specific branch
+  pull_request_target: # Note: `pull_request_target` ensures that the tests run in the context of the `main` branch, not in the user's fork.
+    branches:
+      - dev
+      - main
+      - CAD-22795-restore-changes # Temp
+
+  # Add a manual trigger option for running the workflow
+  workflow_dispatch:
+
+jobs:
+  test:
+    permissions:
+      id-token: write
+      contents: read
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      # Check out the branch from the pull request, whether it's from a fork or not
+      - name: Checkout the pull request's branch
+        run: |
+          set -ex
+          # debug
+          echo "==== before"
+          git status
+          if [[ "${{ github.event_name }}" == "pull_request_target" ]]; then
+            if [ "${{ github.event.pull_request.head.repo.full_name }}" != "${{ github.repository }}" ]; then
+              echo "PR from a fork detected. Checking out the fork's branch."
+              git remote add fork https://github.com/${{ github.event.pull_request.head.repo.full_name }}.git
+              git fetch fork ${{ github.event.pull_request.head.ref }}
+              git checkout -B ci-testing-branch FETCH_HEAD
+            else
+              echo "PR from the same repository detected. Checking out the branch."
+              git fetch origin ${{ github.event.pull_request.head.ref }}
+              git checkout ${{ github.event.pull_request.head.ref }}
+            fi
+          elif [[ "${{ github.event_name }}" == "workflow_dispatch" ]]; then
+            # we believe that `actions/checkout` will checkout to the branch that you chose in the `dispatch` tab
+            # so we do nothing
+            echo "we're in the right branch. no need to checkout (dispatch)"
+          elif [[ "${{ github.event_name }}" == "push" ]]; then
+            # we believe that `actions/checkout` will checkout to the branch that we push into (i.e. either `dev` or `main`)
+            echo "we're in the right branch. no need to checkout (push)"
+          else
+            echo "we're not in push/pull/dispatch. error."
+            exit 1
+          fi
+          # debug
+          echo "==== after"
+          git status
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+
+      # A bunch of if-else. Might move to an action
+      # Decide environment based on the target branch (for both push and PR events)
+      - name: Set environment based on target branch
```

Comment:
Unite with  "Checkout the pull request's branch" step. Its same logic

Reply:
Assure that the workflow still run from target branch!!!!!

Reply:
Add documentation to this step, make sure all your assumptions are documented:
1. what should run on target branch
2. where the tested code shouold come from
3. what is the different cases - give them meaning

Reply:
Steps were united. 
It's still running in the context of the base repo (not target branch), I tried to 

I tried to explain point #3 in the comment at the beginning of the workflow. If it's not clear enough, I would like you to show me what you mean exactly.